scripts/preconditions/preconditions.py

```diff
@@ -0,0 +1,413 @@
+    ]
+
+    # Permissions required when the service account is creating a new project
+    # directly within an organization (as opposed to a folder)
+    PARENT_PERMISSIONS = [
+        # Typically granted with `roles/resourcemanager.projectCreator`
+        "resourcemanager.projects.create"
+    ]
+
+    def __init__(self, org_id, shared_vpc=False, parent=False):
+        """
+        Create a new organization validator.
+
+        Args:
+            org_id (str): The organization ID
+            shared_vpc (bool): Whether shared VPC permissions should be checked
+            parent (bool): Whether parent permissions should be checked
+        """
+        self.org_id = org_id
+        self.shared_vpc = shared_vpc
+        self.parent = parent
+
+        self.permissions = self.ALL_PERMISSIONS[:]
+
+        if self.shared_vpc:
+            self.permissions += self.SHARED_VPC_PERMISSIONS
+
+        if self.parent:
+            self.permissions += self.PARENT_PERMISSIONS
+
+    def validate(self, credentials):
+        service = discovery.build(
+            'cloudresourcemanager', 'v1',
+            credentials=credentials
+        )
+
+        body = {"permissions": self.permissions}
+        resource = "organizations/" + self.org_id
+
+        request = service.organizations().testIamPermissions(
+            resource=resource,
+            body=body)
+        response = request.execute()
+
+        req = Requirements(
+            "Service account permissions on organization",
+            resource,
+            self.permissions,
+            response["permissions"],
+        )
+
+        return req.asdict()
+
+
+class FolderPermissions:
+    # Permissions required when the service account is creating a project
+    # within a folder (as opposed to an organization).
+    PARENT_PERMISSIONS = [
+        # Typically granted with `roles/resourcemanager.projectCreator`
+        "resourcemanager.projects.create",
+
+        # Typically granted with `roles/resourcemanager.folderViewer`
+        "resourcemanager.folders.get",
+    ]
+
+    def __init__(self, folder_id, parent=False):
+        self.folder_id = folder_id
+        self.parent = parent
+        self.permissions = []
+
+        if self.parent:
+            self.permissions += self.PARENT_PERMISSIONS
+
+    def validate(self, credentials):
+        service = discovery.build(
+            'cloudresourcemanager', 'v2',
+            credentials=credentials
+        )
+
+        body = {"permissions": self.permissions}
+        resource = "folders/" + self.folder_id
+
+        request = service.folders().testIamPermissions(
+            resource=resource,
+            body=body)
+        response = request.execute()
+
+        req = Requirements(
+            "Service account permissions on parent folder",
+            resource,
+            self.permissions,
+            response["permissions"],
+        )
+
+        return req.asdict()
+
+
+class HostVpcProjectPermissions:
```

Comment:
Should this be renamed to `SharedVpcProjectPermissions` to match the intended usage?

Reply:
Sure, would you like me to make that change or would you prefer to do it?